control/controlclient/direct.go

```diff
@@ -1303,3 +1303,61 @@ func (c *Direct) SetDNS(ctx context.Context, req *tailcfg.SetDNSRequest) error {
 
 	return nil
 }
+
+// tsmpPing sends a Ping to pr.IP, and sends an http request back to pr.URL
+// with ping response data.
+func tsmpPing(logf logger.Logf, c *http.Client, pr *tailcfg.PingRequest, pinger Pinger) {
+	if pr.URL == "" {
+		logf("invalid PingRequest with no URL")
+		return
+	}
+	if pr.IP.IsZero() {
+		logf("PingRequest with no proper IP got %v", pr.IP)
+		return
+	}
+	if !strings.Contains(pr.Types, "TSMP") {
```

Comment:
if `pr.Types` is a comma separated list, then if `TSMP` is a substring of some other valid entry, this will still match. Better would be to split on commas and then scan the resulting slice for exact match. This works for fine now (with the valid options just being `TSMP` and `disco`), but is a possible surprise later on.

... would it be easier just to make pr.Types a `[]string`? @bradfitz would there be any unfortunate consequences to doing that?